libs/constrained-generators/src/Constrained/List.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Constrained.List where
+
+import Data.Functor.Const
+import Data.Kind
+
+-- | A heterogeneous list / an inductive tuple
+data List (f :: k -> Type) (as :: [k]) where
```

Comment:
Add something like this is the tool we use to define Ctxs
Every Ctx with a hole of type 'a looks something like this
List Value before ++ [Hole a a] ++ List Value after